staking-pool-shares/src/lib.rs

```diff
@@ -0,0 +1,759 @@
+    fn save_account(&mut self, account_id: &AccountId, account: &Account) {
+        if account.unstaked > 0 || account.stake_shares > 0 {
+            self.accounts.insert(&account_id, &account);
+        } else {
+            self.accounts.remove(&account_id);
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use near_sdk::{testing_env, MockedBlockchain};
+
+    use crate::test_utils::*;
+
+    use super::*;
+    use std::convert::TryFrom;
+
+    struct Emulator {
+        pub contract: StakingContract,
+        pub epoch_height: EpochHeight,
+        pub amount: Balance,
+        pub locked_amount: Balance,
+    }
+
+    fn zero_fee() -> RewardFeeFraction {
+        RewardFeeFraction {
+            numerator: 0,
+            denominator: 1,
+        }
+    }
+
+    impl Emulator {
+        pub fn new(
+            owner: String,
+            stake_public_key: String,
+            reward_fee_fraction: RewardFeeFraction,
+        ) -> Self {
+            testing_env!(VMContextBuilder::new()
+                .current_account_id(owner.clone())
+                .finish());
+            Emulator {
+                contract: StakingContract::new(
+                    owner,
+                    Base58PublicKey::try_from(stake_public_key).unwrap(),
+                    reward_fee_fraction,
+                ),
+                epoch_height: 0,
+                amount: 0,
+                locked_amount: 0,
+            }
+        }
+
+        pub fn update_context(&mut self, predecessor_account_id: String, deposit: Balance) {
+            testing_env!(VMContextBuilder::new()
+                .current_account_id(staking())
+                .predecessor_account_id(predecessor_account_id.clone())
+                .signer_account_id(predecessor_account_id)
+                .attached_deposit(deposit)
+                .account_balance(self.amount)
+                .account_locked_balance(self.locked_amount)
+                .epoch_height(self.epoch_height)
+                .finish());
+            println!(
+                "Epoch: {}, Deposit: {}, amount: {}, locked_amount: {}",
+                self.epoch_height, deposit, self.amount, self.locked_amount
+            );
+        }
+
+        pub fn simulate_stake_call(&mut self) {
+            self.update_context(staking(), 0);
+            let total_stake = self.contract.total_staked_balance;
+            // First function call action
+            self.contract.ping();
+            // Stake action
+            self.amount = self.amount + self.locked_amount - total_stake;
+            self.locked_amount = total_stake;
+            // Second function call action
+            self.update_context(staking(), 0);
+            self.contract.internal_after_stake();
+        }
+
+        pub fn skip_epochs(&mut self, num: EpochHeight) {
+            self.epoch_height += num;
+            self.locked_amount = (self.locked_amount * (100 + u128::from(num))) / 100;
```

Comment:
Please, comment these numbers

Reply:
Is that correct that node re-invest staking reward back to the locked_amount automatically?

Reply:
And the reward per epoch is 1%, correct?

Reply:
These are fake reward numbers